src/datasets/formatting/formatting.py

```diff
@@ -187,14 +187,20 @@ def _arrow_array_to_numpy(self, pa_array: pa.Array) -> np.ndarray:
             else:
                 zero_copy_only = _is_zero_copy_only(pa_array.type) and not _is_array_with_nulls(pa_array)
                 array: List = pa_array.to_numpy(zero_copy_only=zero_copy_only).tolist()
+
         if len(array) > 0:
             if any(
                 (isinstance(x, np.ndarray) and (x.dtype == object or x.shape != array[0].shape))
                 or (isinstance(x, float) and np.isnan(x))
                 for x in array
             ):
+                if np.lib.NumpyVersion(np.__version__) >= '2.0.0b1':
+                    return np.array(array, dtype=object)
                 return np.array(array, copy=False, dtype=object)
-        return np.array(array, copy=False)
+        if np.lib.NumpyVersion(np.__version__) >= '2.0.0b1':
+            return np.array(array)
```

Comment:
I think it should be `asarray` to avoid copies when possible no ?
```suggestion
            return np.asarray(array)
```

Reply:
You are completely right

Reply:
I believe we can actually remove the if else statement and just use asarray() with older versions of numpy as well. That is at least my reading of the [docs](https://numpy.org/doc/1.26/reference/generated/numpy.asarray.html)

Reply:
they state that asarray used to have more overhead than copy=False. So maybe keep copy=False for numpy 1.x and use asarray for 2.x